

```{r eval=FALSE}

library(harmony)
library(Seurat)
library(tidyverse)
library(cowplot)
library(patchwork)
library(WGCNA)
library(Matrix)
library(viridis)
library(RColorBrewer)
library(ggpubr)
library(tictoc)
library(RColorBrewer)
library(Hmisc)
library(enrichR)
library(GeneOverlap)
library(grid)
library(gridExtra)
library(igraph)
library(ggrepel)

#library(hdWGCNA)
enableWGCNAThreads(nThreads = 8)
theme_set(theme_cowplot())
set.seed(123456)

library(hdWGCNA)

setwd("/dfs7/swaruplab/zechuas/Collaborations/MODEL_AD/Bin1_5xFAD_snRNAseq_Nov2022/Analysis/hdWGCNA/ODC_COP/")
data_dir <- 'data/'
fig_dir <- 'figures/'

dir.create(data_dir)
dir.create(fig_dir)


seurat_PATH <- '/dfs7/swaruplab/zechuas/Collaborations/MODEL_AD/Bin1_5xFAD_snRNAseq_Nov2022/Analysis/Subcluster_analysis/'
seurat_obj <- readRDS(paste0(seurat_PATH, 'ODC_COP/', 'NucSeq_ODC_clusters_ODC_Status_Updates_06242023.rds'))


```

Run hdWGCNA on one cell type (ODC)

```{r eval=FALSE}

# seurat_subset <- seurat_obj
seurat_subset <- seurat_obj %>% subset(ODC_Status %in% c('DAO', 'Homeostatic MOL', 'MOL'))
table(seurat_subset@meta.data$ODC_Status)

seurat_subset <- SetupForWGCNA(
  seurat_subset,
  gene_select = "fraction",
  fraction = 0.05,
  group.by='ODC_Status',
  wgcna_name = 'ODC'
)



seurat_subset<- MetacellsByGroups(
  seurat_obj = seurat_subset,
  group.by = c('Genotype', 'ODC_Status'),
  k = 50,
  target_metacells=250,
  ident.group = 'ODC_Status',
  min_cells=100,
  max_shared=20,
)

seurat_subset <- NormalizeMetacells(seurat_subset)

seurat_subset <- SetDatExpr(
  seurat_subset,
  group.by= 'ODC_Status',
  group_name = c('DAO', 'Homeostatic MOL', 'MOL'),
  assay = 'RNA'
)

# test different values for soft power
seurat_subset <- TestSoftPowers(seurat_subset)

# plot the results:
plot_list <- PlotSoftPowers(seurat_subset)

pdf(paste0(fig_dir, 'test_softpower_ODC.pdf'), width=12, height=9)
wrap_plots(plot_list, ncol=2)
dev.off()

# NOTE
seurat_subset <- ConstructNetwork(
    seurat_subset,
    tom_name='ODC',
    minModuleSize= 30, # minModuleSize=30,
    detectCutHeight=0.995,
    mergeCutHeight=0.05,
    overwrite_tom=TRUE
)

# plot the dendrogram
pdf(paste0(fig_dir, "dendro_ODC.pdf"),height=3, width=6)
PlotDendrogram(seurat_subset, main='ODC hdWGCNA Dendrogram')
dev.off()

# compute module eigengenes and apply harmony correction by Sample
seurat_subset <- ModuleEigengenes(
  seurat_subset,
  group.by.vars = 'sample',
  verbose=TRUE
)

# compute eigengene-based connecivity (kME)
# must specify the same groups that you did with SetDatExpr
seurat_subset <- ModuleConnectivity(
    seurat_subset,
    group.by= 'ODC_Status',
    group_name = c('DAO', 'Homeostatic MOL', 'MOL')
)

MEs <- GetMEs(seurat_subset)
modules <- GetModules(seurat_subset)
mods <- levels(modules$module)
mods <- mods[mods!='grey']

mod_colors_df <- dplyr::select(modules, c(module, color)) %>%
   distinct %>% arrange(module)
mod_colors_df
# > mod_colors_df


meta <- seurat_subset@meta.data
seurat_subset@meta.data <- cbind(meta, MEs)

# make dotplot
p <- DotPlot(
  seurat_subset,
  group.by='ODC_Status', # group.by='Celltypes',
  features = rev(mods)
) + RotatedAxis() +
  scale_color_gradient2(high='red', mid='grey95', low='blue') + xlab('') + ylab('') +
  theme(
    plot.title = element_text(hjust = 0.5),
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    panel.border = element_rect(colour = "black", fill=NA, size=1)
  )

pdf(paste0(fig_dir, 'dotplot_MEs_ODC.pdf'), width=8, height=4)
p
dev.off()


seurat_subset@meta.data <- meta


```

Reset the module names and colors

```{r eval=FALSE}

# rename the modules
seurat_subset <- ResetModuleNames(
  seurat_subset,
  new_name = "ODC-M"
)

modules <- GetModules(seurat_subset)
mods <- levels(modules$module)
#

mod_colors_df <- dplyr::select(modules, c(module, color)) %>%
   distinct %>% arrange(module)
mod_colors_df
# > mod_colors_df

write.csv(modules, quote=FALSE, row.names=FALSE, file=paste0(data_dir, 'ODC_modules.csv'))
write.csv(mod_colors_df, quote=FALSE, row.names=FALSE, file=paste0(data_dir, 'ODC_modules_colors_df.csv'))


```


```{r eval=FALSE}

p <- PlotKMEs(seurat_subset, ncol=4)

pdf(paste0(fig_dir, 'ODC_kME_distributions.pdf'), width=15, height=12)
p
dev.off()

modules <- GetModules(seurat_subset)

```



Network plots
```{r eval=FALSE}

# individual module network circle plots
ModuleNetworkPlot(
  seurat_subset,
  mods = "all",
  #label_center=TRUE,
  outdir = paste0(fig_dir, 'ODC_hubNetworks/')
)


seurat_subset <- RunModuleUMAP(
    seurat_subset,
    n_hubs =5,
    n_neighbors=25,
    min_dist=0.5,
    spread=1.5,
    supervised=TRUE,
    target_weight=0.2
)


# get the hub gene UMAP table from the seurat object
umap_df <- GetModuleUMAP(
  seurat_subset
)

# plot with ggplot
p <- ggplot(umap_df, aes(x=UMAP1, y=UMAP2)) +
  geom_point(
   color=umap_df$color,
   size=umap_df$kME*2
  ) +
  umap_theme()

pdf(paste0(fig_dir, 'test_hubgene_umap_ggplot_uns.pdf'), width=5, height=5)
p
dev.off()



```

Run enrichr on the modules

```{r eval=FALSE}

library(enrichR)

dbs <-c('GO_Biological_Process_2021','GO_Cellular_Component_2021','GO_Molecular_Function_2021', 'WikiPathway_2021_Mouse', 'KEGG_2021_Mouse')

# compute GO terms:
# set the number of genes to test per module or set max_genes=Inf to include all genes
seurat_subset <- RunEnrichr(seurat_subset, dbs=dbs, max_genes=Inf)

# subset the table for significant results and save
# inspect this table in Excel to select relevant terms for each module
enrichr_df <- GetEnrichrTable(seurat_subset) %>% subset(P.value < 0.05)
write.table(enrichr_df, quote=FALSE, sep='\t', row.names=FALSE, file=paste0(data_dir, 'ODC_enrichr.tsv'))
# write.csv(enrichr_df, quote=FALSE, row.names=FALSE, file=paste0(data_dir, 'MG_enrichr.csv'))
save(enrichr_df, file=paste0(data_dir, 'ODC_enrichr.rda'))


```

Save

```{r eval=FALSE}

saveRDS(seurat_subset, file=paste0(data_dir, 'NucSeq_ODC_status_06252023_hdWGCNA.rds'))


```

Differential module eigengene (DME) analysis

```{r eval=FALSE}

# list of genotypes for pairwise comparisons
gt1_list <- c('5xFAD_HEMI', '5xFAD_HEMI-BIN1HO', '5xFAD_HEMI-BIN1HO', '5xFAD_HEMI-BIN1HO', "BIN1HO")
gt2_list <- c("5xFAD_WT", "BIN1HO", '5xFAD_HEMI', "5xFAD_WT", "5xFAD_WT")

######################################################
# 5xFAD Hemi vs 5xFAD WT
######################################################

DMEs <- Reduce(rbind, lapply(1:length(gt1_list), function(i){

    # get current genotypes to compare
    gt1 <- gt1_list[i]
    gt2 <- gt2_list[i]

    # get list of cells with these genotypes
    g1 <- seurat_subset@meta.data %>% subset(Genotype == gt1) %>% rownames
    g2 <- seurat_subset@meta.data %>% subset(Genotype == gt2) %>% rownames

    # run the DME comparison with a wilcox test
    DMEs <- FindDMEs(
        seurat_subset,
        barcodes1 = g1,
        barcodes2 = g2,
        test.use='wilcox',
        wgcna_name='ODC'
    )
    DMEs$ident.1 <- gt1
    DMEs$ident.2 <- gt2
    DMEs$comparison <- paste0(gt1, '_vs_', gt2)
    DMEs
}))

# compute FDR across all:
DMEs$fdr <- p.adjust(DMEs$p_val, method='fdr')

################################################################################
# Plot DMEs as a lollipop
################################################################################
# devtools::install_github("NightingaleHealth/ggforestplot")

comparisons <- unique(DMEs$comparison)
plot_list <- list()
for(cur_comp in comparisons){

    print(cur_comp)

    cur_DMEs <- subset(DMEs, comparison == cur_comp)
    gt1 <- unique(cur_DMEs$ident.1)
    gt2 <- unique(cur_DMEs$ident.2)
    cur_title <- paste0(gt1, ' vs. ', gt2)

    # set plotting attributes for shape
    cur_DMEs$shape <- ifelse(cur_DMEs$fdr < 0.05, 21,4)
    cur_DMEs <- cur_DMEs %>% arrange(avg_log2FC, descending=TRUE)
    cur_DMEs$module <- factor(as.character(cur_DMEs$module), levels=as.character(cur_DMEs$module))

    # add number of genes per module
    n_genes <- table(modules$module)
    cur_DMEs$n_genes <- as.numeric(n_genes[as.character(cur_DMEs$module)])

    mod_colors <- dplyr::select(modules, c(module, color)) %>%
    distinct
    cp <- mod_colors$color; names(cp) <- mod_colors$module

    p <- cur_DMEs %>%
    ggplot(aes(y=module, x=avg_log2FC, size=log(n_genes), color=module)) +
    geom_vline(xintercept=0, color='black') +
    geom_segment(aes(y=module, yend=module, x=0, xend=avg_log2FC), linewidth=0.5, alpha=0.3) +
    geom_point() +
    geom_point(shape=cur_DMEs$shape, color='black', fill=NA) +
    scale_color_manual(values=cp, guide='none') +
    ylab('') +
    xlab(bquote("Avg. log"[2]~"(Fold Change)")) +
    theme(
        axis.line.y = element_blank(),
        axis.ticks.y = element_blank(),
        plot.title = element_text(hjust=0.5, face='plain', size=10)
    ) +
    ggtitle(cur_title) + NoLegend()

    p <- p +  ggforestplot::geom_stripes(aes(y=module), inherit.aes=FALSE, data=cur_DMEs)

    plot_list[[cur_comp]] <- p

}

# individual plots

# pdf(paste0(fig_dir, 'MG_DMEs_lollipop.pdf'), width=4, height=3.5)
pdf(paste0(fig_dir, 'ODC_DMEs_lollipop.pdf'), width=4, height=3.5)
for(p in plot_list){
    print(p)
}
dev.off()

# combined plots

# pdf(paste0(fig_dir, 'MG_DMEs_lollipop_combined.pdf'), width=12, height=7.5)
pdf(paste0(fig_dir, 'ODC_DMEs_lollipop_combined.pdf'), width=15, height=11)
wrap_plots(plot_list, ncol=3)
dev.off()
